zeppelin-user-impersonation[3016].md

```diff
@@ -0,0 +1,152 @@
+#/api/credential/** = authc, roles[admin] </br>
+#/** = anon </br>
+/** = authc </br>
+</i>
+
+<b>Below are the configurations required from the environment:</b> </br>
+•	On the Zeppelin server node make sure to add the below to <i>/etc/sudoers</i>: </br>
+With the root user: </br>
+
+`$visudo` </br>
+`zeppelin        ALL=(ALL)       NOPASSWD: ALL ` </br> 
+
+If this entry is missing in the /etc/sudoers then the interpreter fails with the below error: </br>
+
+`org.apache.zeppelin.interpreter.InterpreterException: Permission denied, please try again.`</br>
+`Permission denied, please try again.` </br>
+`Permission denied (publickey,gssapi-keyex,gssapi-with-mic,password).`</br> 
+</br>
+
+•	Log folder (usually /var/log/zeppelin or /var/log/hadoop/zeppelin) of zeppelin should be of 777 instead of 755 if not the Spark interpreter keep failing with below error as the impersonated user won’t able to write to the logs folder:
+
+`java.lang.NullPointerException`</br>
+	`at org.apache.zeppelin.spark.Utils.invokeMethod(Utils.java:38)`</br>
+	`at org.apache.zeppelin.spark.Utils.invokeMethod(Utils.java:33)` </br>
+
+</br>
+•	If you are on a HDP platform this can’t be achieved from the CLI which needs a Ambari Zeppelin code level change </br>
+
+`$ vi /var/lib/ambari-server/resources/common-services/ZEPPELIN/0.6.0.2.5/package/scripts/master.py` </br>
+
+#### Find the below: </br>
+
+```
+def create_zeppelin_log_dir(self, env):
+    import params
+    env.set_params(params)
+    Directory([params.zeppelin_log_dir],
+              owner=params.zeppelin_user,
+              group=params.zeppelin_group,
+              cd_access="a",
+              create_parents=True,
+              mode=0755
+              )
+  ```          
+##### Change this to:
+
+```
+def create_zeppelin_log_dir(self, env):
+    import params
+    env.set_params(params)
+    Directory([params.zeppelin_log_dir],
+              owner=params.zeppelin_user,
+              group=params.zeppelin_group,
+              cd_access="a",
+              create_parents=True,
+              mode=0777
+              )
+```
+##### Please note this requires the stop and start of AMBARI-SERVER. </br>
+Configurations required for Zeppelin User Impersonation: </br>
+
+•	Add the proxy settings for zeppelin user under core-site.xml and restart the required services (HDFS, Mapreduce2 and YARN) </br>
+
+`hadoop.proxyuser.zeppelin.hosts=*` </br>
+`hadoop.procyuser.zeppelin.goups=*`
+</br>
+
+This configuration is required as we will be giving the control to zeppelin user to impersonate other users (users using the zeppelin UI) </br>
+
+•	Add the below lines to zeppelin-env.sh file (or zeppelin-env from Ambari --> Zeppelin --> Configs --> Advanced zeppelin-env)
+
+<i>#--- FOR IMPERSONATION -- START ----
+
+ZEPPELIN_IMPERSONATE_USER=`echo ${ZEPPELIN_IMPERSONATE_USER} | cut -d "@" -f1` # we are trimming the user to get short name i.e. username from username@domain.com
```

Comment:
This is HDP specific, not for apache zeppelin. And are you sure zeppelin in HDP needs that ? because from my experience, I don't need to set these to make impersonation work in mac at least.

Reply:
Yes, after testing the functionality with out these, then i had to come up with this workaround

Reply:
I opened a PR (https://github.com/apache/zeppelin/pull/2719) that solves the issue of user impersonation on YARN (using Spark stock version)  or Mesos (using Mesosphere specific Spark fork) without the need to customize DHFS distro or Zeppelin env vars.